File: typescript/0002-Add-Two-Numbers/add-two-numbers.ts
import { ListNode, IListNode } from "../data-structures/list-node";

export function AddTwoNumbers(
    list1 : IListNode<number> | null, 
    list2: IListNode<number> | null
    ) : IListNode<number> | null {
        const holder = new ListNode(0);
        let rtn = holder;
        let carry : number = 0;

        while (list1 || list2 || carry) {
            if (list1) {
                carry += list1.val;
                list1 = list1.next;
            }

            if (list2) {
                carry += list2.val;
                list2 = list2.next;
            }

            if (rtn != null) {
            rtn.next = new ListNode(carry % 10);
            rtn = rtn.next;
            }
            carry = Math.floor(carry / 10);
        }

    return holder.next;
}

File: go/0002-Add-Two-Numbers/addtwonumbers.go

package leetcode

type ListNode struct {
    Val int
    Next *ListNode
}

func AddTwoNumbers(input1 *ListNode, input2 *ListNode) *ListNode {
    var rtn *ListNode = &ListNode{ Val: 0 }
    var x, y, carry, sum int = 0, 0, 0, 0
    
    node := rtn
    ln1 := input1
    ln2 := input2

    for ln1 != nil || ln2 != nil || carry != 0 {
        if (ln1 != nil) { 
            x = ln1.Val 
            ln1 = ln1.Next
        }
        if (ln2 != nil) { 
            y = ln2.Val 
            ln2 = ln2.Next
        }

        sum = x + y + carry
        carry = sum / 10
        node.Next = &ListNode { Val: sum /10 } 
        
        node = node.Next 
        if (ln1 != nil) {ln1 = ln1.Next}
        if (ln2 != nil) {ln2 = ln2.Next}
    }

    return rtn
}
